src/server/game/Entities/Player/Player.cpp

```diff
@@ -28727,3 +28731,74 @@ uint8 Player::GetItemLimitCategoryQuantity(ItemLimitCategoryEntry const* limitEn
 
     return limit;
 }
+
+bool Player::IsAtMaxLevel() const
+{
+    return getLevel() >= sWorld->getIntConfig(CONFIG_MAX_PLAYER_LEVEL);
+}
+
+void Player::SetWarModeDesired(bool enabled)
+{
+    // Only allow to toggle on when in stormwind/orgrimmar, and to toggle off in any rested place.
+    // Also disallow when in combat
+    if ((enabled == IsWarModeDesired()) || IsInCombat() || !IsInRestArea())
+        return;
+
+    if (enabled && (IsTeamAlliance() ? (GetZoneId() != ZONE_STORMWIND_CITY) : (GetZoneId() != ZONE_ORGRIMMAR)))
+        return;
+
+    if (enabled)
+    {
+        AddPlayerFlag(PLAYER_FLAGS_WAR_MODE_DESIRED);
+        TogglePvpTalents(true);
+        SetPvP(true);
+        RemoveUnitFlag(UNIT_FLAG_PVP_ATTACKABLE);
+    }
+    else
+    {
+        RemovePlayerFlag(PLAYER_FLAGS_WAR_MODE_DESIRED);
+        TogglePvpTalents(false);
+        SetPvP(false);
+        AddUnitFlag(UNIT_FLAG_PVP_ATTACKABLE);
+    }
+
+    UpdateWarModeAuras();
+}
+
+void Player::UpdateWarModeAuras()
+{
+    uint32 auraInside = 282559;
+    uint32 auraOutside = WARMODE_ENLISTED_SPELL_OUTSIDE;
+    uint32 auraOutsideMaxLvl = 289954;
+
+    if (IsWarModeDesired())
+    {
+        bool moveInsideCity = IsTeamAlliance() ? (m_zoneId == ZONE_STORMWIND_CITY) : (m_zoneId == ZONE_ORGRIMMAR);
+        if (moveInsideCity)
+        {
+            RemoveAurasDueToSpell(auraOutside);
+            RemoveAurasDueToSpell(auraOutsideMaxLvl);
+            CastSpell(this, auraInside, true);
+            RemovePlayerFlag(PLAYER_FLAGS_WAR_MODE_ACTIVE);
+        }
+        else
+        {
+            RemoveAurasDueToSpell(auraInside);
+
+            TeamId team = sWorld->GetCurrentFactionBalanceTeam();
+            CustomSpellValues const* values = nullptr;
+            if (GetTeamId() == team)
+                values = &(sWorld->GetCurrentFactionBalanceRewardSpellValues());
+            CastSpell(this, IsAtMaxLevel() ? auraOutsideMaxLvl : auraOutside, true, nullptr, nullptr, ObjectGuid::Empty, values);
```

Comment:
Use `void Unit::CastCustomSpell(uint32 spellId, CustomSpellValues const& value, Unit* victim, TriggerCastFlags triggerFlags, Item* castItem, AuraEffect const* triggeredByAura, ObjectGuid originalCaster)`

Reply:
Done.

There is a `CastSpell` override that accepts `CustomSpellValues` as one of it's argument; should it be renamed to CastCustomSpell ?